cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +308,435 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+      printf("%3d ", nodeII.element(n, stream));
+    printf(" nodeII-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", parent_node_idx.element(n, stream));
+    printf(" parent_node_idx-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", parent_col_id.element(n, stream));
+    printf(" parent_col_id-%ld\n", level);
+    for (auto n = start; n < end; n++) {
+      auto nt = node_type.element(n, stream);
+      printf("%3s ",
+             nt >= NUM_NODE_CLASSES ? std::to_string(nt - NUM_NODE_CLASSES).c_str() : cls[nt]);
+    }
+    printf(" node_type-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", levels.element(n, stream));
+    printf(" levels-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", col_id.element(n, stream));
+    printf(" col_id-%ld\n", level);
+  };
+#define PRINT_LEVEL_DATA(level, text)           \
+  print_level_data(text,                        \
+                   level,                       \
+                   level_boundaries[level - 1], \
+                   level_boundaries[level],     \
+                   scatter_indices,             \
+                   parent_indices,              \
+                   parent_col_id,               \
+                   node_type,                   \
+                   d_tree.node_levels,          \
+                   col_id);
+#else
+#define PRINT_LEVEL_DATA(level, text) ;
+#endif
+
+  // 4. Propagate parent node ids for each level.
+  // For each level,
+  //     a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+  //     b. stable sort by {parent_col_id, node_type}
+  //     c. scan sum of unique {parent_col_id, node_type}
+  //     d. scatter the col_id back to stable node_level order (using scatter_indices)
+  nvtxRangePushA("pre-level");
+  rmm::device_uvector<NodeIndexT> col_id(num_nodes, stream, mr);
+  rmm::device_uvector<NodeIndexT> parent_col_id(num_nodes, stream);
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
+                             parent_col_id.begin(),
+                             parent_col_id.end(),
+                             0);  // XXX: is this needed?
+  // fill with 1, useful for scan later
+  // TODO: Could initialize to 0 and scatter 1 to level_boundaries alone
+  thrust::uninitialized_fill(rmm::exec_policy(stream), col_id.begin(), col_id.end(), 1);
+  // Initialize First level node's node col_id to 0
+  thrust::fill(rmm::exec_policy(stream), col_id.begin(), col_id.begin() + level_boundaries[0], 0);
+  // Initialize First level node's parent_col_id to parent_node_sentinel sentinel
+  thrust::fill(rmm::exec_policy(stream),
+               parent_col_id.begin(),
+               parent_col_id.begin() + level_boundaries[0],
+               parent_node_sentinel);
+  nvtxRangePop();
+  nvtxRangePushA("level");
+  for (decltype(num_levels) level = 1; level < num_levels; level++) {
```

Comment:
Leaving a sample debug print here as support for other reviewers to understand the process:

```
num_levels: 6
1.before gather
  2   9  20  28  34  nodeII-1
  1   2   3   4   4  parent_node_idx-1
  0   0   0   0   0  parent_col_id-1
 20  20  20  20  34  node_type-1
  1   1   1   1   1  levels-1
  1  39  38   4  15  col_id-1
1.after gather
  2   9  20  28  34  nodeII-1
  1   2   3   4   4  parent_node_idx-1
  0   0   0   0   0  parent_col_id-1
 20  20  20  20  34  node_type-1
  1   1   1   1   1  levels-1
  1  39  38   4  15  col_id-1
1.after sort
  2   9  20  28  34  nodeII-1
  1   2   3   4   4  parent_node_idx-1
  0   0   0   0   0  parent_col_id-1
 20  20  20  20  34  node_type-1
  1   1   1   1   1  levels-1
  1  39  38   4  15  col_id-1
1.after scan
  2   9  20  28  34  nodeII-1
  1   2   3   4   4  parent_node_idx-1
  0   0   0   0   0  parent_col_id-1
 20  20  20  20  34  node_type-1
  1   1   1   1   1  levels-1
  1   1   1   1   2  col_id-1
1.after restore order
  2   9  20  28  34  nodeII-1
  1   2   3   4   4  parent_node_idx-1
  0   0   0   0   0  parent_col_id-1
 20  20  20  20  34  node_type-1
  1   1   1   1   1  levels-1
  1   1   1   1   2  col_id-1
```

<details>
  <summary>Debug prints for all levels</summary>

```
2.before gather
  3  10  21  29  35  nodeII-2
  5   6   7   8   9  parent_node_idx-2
  0   0   0   0   0  parent_col_id-2
  S   S   S   S   S  node_type-2
  2   2   2   2   2  levels-2
 16   4  16  14   4  col_id-2
2.after gather
  3  10  21  29  35  nodeII-2
  5   6   7   8   9  parent_node_idx-2
  1   1   1   1   2  parent_col_id-2
  S   S   S   S   S  node_type-2
  2   2   2   2   2  levels-2
 16   4  16  14   4  col_id-2
2.after sort
  3  10  21  29  35  nodeII-2
  5   6   7   8   9  parent_node_idx-2
  1   1   1   1   2  parent_col_id-2
  S   S   S   S   S  node_type-2
  2   2   2   2   2  levels-2
 16   4  16  14   4  col_id-2
2.after scan
  3  10  21  29  35  nodeII-2
  5   6   7   8   9  parent_node_idx-2
  1   1   1   1   2  parent_col_id-2
  S   S   S   S   S  node_type-2
  2   2   2   2   2  levels-2
 16  16  16  16  17  col_id-2
2.after restore order
  3  10  21  29  35  nodeII-2
  5   6   7   8   9  parent_node_idx-2
  1   1   1   1   2  parent_col_id-2
  S   S   S   S   S  node_type-2
  2   2   2   2   2  levels-2
 16  16  16  16  17  col_id-2
3.before gather
  4   6  11  13  22  25  30  32  36  38  nodeII-3
 10  10  11  11  12  12  13  13  14  14  parent_node_idx-3
  0   0   0   0   0   0   0   0   0   0  parent_col_id-3
 32  38  32  38  38  32  36  32  36  38  node_type-3
  3   3   3   3   3   3   3   3   3   3  levels-3
 34  14   4  18  14   4  24  23   4  50  col_id-3
3.after gather
  4   6  11  13  22  25  30  32  36  38  nodeII-3
 10  10  11  11  12  12  13  13  14  14  parent_node_idx-3
 16  16  16  16  16  16  16  16  17  17  parent_col_id-3
 32  38  32  38  38  32  36  32  36  38  node_type-3
  3   3   3   3   3   3   3   3   3   3  levels-3
 34  14   4  18  14   4  24  23   4  50  col_id-3
3.after sort
  4  11  25  32  30   6  13  22  36  38  nodeII-3
 10  10  11  11  12  12  13  13  14  14  parent_node_idx-3
 16  16  16  16  16  16  16  16  17  17  parent_col_id-3
 32  32  32  32  36  38  38  38  36  38  node_type-3
  3   3   3   3   3   3   3   3   3   3  levels-3
 34  14   4  18  14   4  24  23   4  50  col_id-3
3.after scan
  4  11  25  32  30   6  13  22  36  38  nodeII-3
 10  10  11  11  12  12  13  13  14  14  parent_node_idx-3
 16  16  16  16  16  16  16  16  17  17  parent_col_id-3
 32  32  32  32  36  38  38  38  36  38  node_type-3
  3   3   3   3   3   3   3   3   3   3  levels-3
 34  34  34  34  35  36  36  36  37  38  col_id-3
3.after restore order
  4   6  11  13  22  25  30  32  36  38  nodeII-3
 10  10  11  11  12  12  13  13  14  14  parent_node_idx-3
 16  16  16  16  16  16  16  16  17  17  parent_col_id-3
 32  32  32  32  36  38  38  38  36  38  node_type-3
  3   3   3   3   3   3   3   3   3   3  levels-3
 34  36  34  36  36  34  35  34  37  38  col_id-3
4.before gather
  5   7  12  14  23  26  31  33  37  39  nodeII-4
 15  16  17  18  19  20  21  22  23  24  parent_node_idx-4
  0   0   0   0   0   0   0   0   0   0  parent_col_id-4
  V   L   V   L   L   V   V   V   V   V  node_type-4
  4   4   4   4   4   4   4   4   4   4  levels-4
 89  -1  52  -1  71  72  73  -1  74  -1  col_id-4
4.after gather
  5   7  12  14  23  26  31  33  37  39  nodeII-4
 15  16  17  18  19  20  21  22  23  24  parent_node_idx-4
 34  36  34  36  36  34  35  34  37  38  parent_col_id-4
  V   L   V   L   L   V   V   V   V   V  node_type-4
  4   4   4   4   4   4   4   4   4   4  levels-4
 89  -1  52  -1  71  72  73  -1  74  -1  col_id-4
4.after sort
  5  12  26  33  31   7  14  23  37  39  nodeII-4
 15  16  17  18  19  20  21  22  23  24  parent_node_idx-4
 34  34  34  34  35  36  36  36  37  38  parent_col_id-4
  V   V   V   V   V   L   L   L   V   V  node_type-4
  4   4   4   4   4   4   4   4   4   4  levels-4
 89  -1  52  -1  71  72  73  -1  74  -1  col_id-4
4.after scan
  5  12  26  33  31   7  14  23  37  39  nodeII-4
 15  16  17  18  19  20  21  22  23  24  parent_node_idx-4
 34  34  34  34  35  36  36  36  37  38  parent_col_id-4
  V   V   V   V   V   L   L   L   V   V  node_type-4
  4   4   4   4   4   4   4   4   4   4  levels-4
 89  89  89  89  90  91  91  91  92  93  col_id-4
4.after restore order
  5   7  12  14  23  26  31  33  37  39  nodeII-4
 15  16  17  18  19  20  21  22  23  24  parent_node_idx-4
 34  36  34  36  36  34  35  34  37  38  parent_col_id-4
  V   V   V   V   V   L   L   L   V   V  node_type-4
  4   4   4   4   4   4   4   4   4   4  levels-4
 89  91  89  91  91  89  90  89  92  93  col_id-4
5.before gather
 15  16  17  18  24  nodeII-5
 28  28  28  28  29  parent_node_idx-5
  0   0   0   0   0  parent_col_id-5
  V   V   V   V   V  node_type-5
  5   5   5   5   5  levels-5
184  -1  92  -1 109  col_id-5
5.after gather
 15  16  17  18  24  nodeII-5
 28  28  28  28  29  parent_node_idx-5
 91  91  91  91  91  parent_col_id-5
  V   V   V   V   V  node_type-5
  5   5   5   5   5  levels-5
184  -1  92  -1 109  col_id-5
5.after sort
 15  16  17  18  24  nodeII-5
 28  28  28  28  29  parent_node_idx-5
 91  91  91  91  91  parent_col_id-5
  V   V   V   V   V  node_type-5
  5   5   5   5   5  levels-5
184  -1  92  -1 109  col_id-5
5.after scan
 15  16  17  18  24  nodeII-5
 28  28  28  28  29  parent_node_idx-5
 91  91  91  91  91  parent_col_id-5
  V   V   V   V   V  node_type-5
  5   5   5   5   5  levels-5
184 184 184 184 184  col_id-5
5.after restore order
 15  16  17  18  24  nodeII-5
 28  28  28  28  29  parent_node_idx-5
 91  91  91  91  91  parent_col_id-5
  V   V   V   V   V  node_type-5
  5   5   5   5   5  levels-5
184 184 184 184 184  col_id-5
``` 
</details>

Reply:
Does it make sense to call this an unrolled BFS traversal? Maybe that also helps conceptually

Reply:
```

#ifdef NJP_DEBUG_PRINT
  auto print_level_data = [stream](auto stage_text,
                                   auto level,
                                   auto start,
                                   auto end,
                                   auto const& nodeII,
                                   auto const& parent_node_idx,
                                   auto const& parent_col_id,
                                   auto const& node_type,
                                   auto const& col_id) {
    std::cout << level << stage_text << "\n";
    auto cls  = std::array<const char*, 6>{"S", "L", "F", "R", "V", "E"};
    auto dget = [](auto const& v, auto i) { return int(thrust::device_pointer_cast(v.data())[i]); };
    for (auto n = start; n < end; n++)
      printf("%3d ", dget(nodeII, n));
    printf(" nodeII-%ld\n", level);
    for (auto n = start; n < end; n++)
      printf("%3d ", dget(parent_node_idx, n));
    printf(" parent_node_idx-%ld\n", level);
    for (auto n = start; n < end; n++)
      printf("%3d ", parent_col_id.element(n, stream));
    printf(" parent_col_id-%ld\n", level);
    for (auto n = start; n < end; n++) {
      auto nt = dget(node_type, n);
      printf("%3s ",
             nt >= NUM_NODE_CLASSES ? std::to_string(nt - NUM_NODE_CLASSES).c_str() : cls[nt]);
    }
    printf(" node_type-%ld\n", level);
    for (auto n = start; n < end; n++)
      printf("%3d ", col_id.element(n, stream));
    printf(" col_id-%ld\n", level);
  };
#define PRINT_LEVEL_DATA(level, text)           \
  print_level_data(text,                        \
                   level,                       \
                   level_boundaries[level - 1], \
                   level_boundaries[level],     \
                   scatter_indices,             \
                   parent_indices,              \
                   parent_col_id,               \
                   node_type,                   \
                   col_id);
#else
#define PRINT_LEVEL_DATA(level, text) ;
#endif
```
keeping this snippet here for future debug purposes.